text/3668-async-closure.md

```diff
@@ -0,0 +1,713 @@
+
+### Why do we recommend the `AsyncFnOnce::Output` type remains unstable, unlike `FnOnce::Output`?
+
+As mentioned above, `FnOnce::Output` was stabilized in [#34365](https://github.com/rust-lang/rust/pull/34365) as an alternative to break ecosystem code when a bug was fixed to detect usage of unstable associated items in "type-dependent" associated type paths (e.g. `T::Output` that is not qualified with a trait). This allows the following code on stable:
+
+```rust
+fn foo<F, T>()
+where
+    F: FnOnce() -> T,
+    F::Output: Send, //~ OK
+{
+}
+```
+
+However, the stabilization of the assoicated type did not actually enable new things to be expressed, and instead `FnOnce::Output` just serves as a type alias for an existing type that may already be named.
+
+This is because uniquely to `Fn*` trait bounds (compared to the other `std::ops::*` traits that define `Output` associated types, like `Add`), the associated type for `FnOnce::Output` is always constrained by the parenthesized generic syntax. In other words, given `F: Fn*() -> T`, `F::Output` can always be replaced by some type `T`, since `T` is necessary to complete the parenthesized trait bound syntax[^higher]. In that way, naming a type via the `Output` associated type is not more general or flexible than just naming the type itself:
+
+[^higher]: In fact, the `::Output` syntax doesn't even make it easier to name the return type of a higher-ranked `Fn` trait bound either: https://godbolt.org/z/1rTGhfr9x
+
+```rust
+fn foo<F, T>()
+where
+    F: FnOnce() -> T,
+    F::Output: Send,
+    // Should just be rewritten like:
+    T: Send,
+{
+}
+```
+
+Exposing the `Output` type for `AsyncFnOnce` complicates eventually moving onto other desugarings for `async Fn*`. For example, if `AsyncFnOnce` is replaced by a trait alias for `FnOnce`, it may change the meaning of `Output` in a way that would require extending the language or adding a hack into the compiler to preserve its meaning.
+
+Given that expressivity isn't meaningfully impaired by keeping the `Output` associated type as unstable, we do not expect to stabilize this associated type at the same time as async closures, and a stabilization report for the associated type should mention how it affects future possibilities to change the desugaring of `async Fn*`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+### Users might confusedly write `|| async {}` over `async || {}`
+
+Users may be confused whether to write `|| async {}` or `async || {}`. The fact that `async || {}` has extra "superpowers" with respect to lending may lead to users hitting unnecessary errors if they invert the ordering.
+
+We should be able to detect when users write `|| async {}` -- and subsequently hit borrow checker issues -- and give a useful error message to move the `async` keyword. We may also lint against `|| async {}` in code that *does* pass, since it's not as expressive.
```

Comment:
```suggestion
We should be able to detect when users write `|| async {}` -- and subsequently hit borrow checker issues -- and give a useful error message to move the `async` keyword. We may also lint against `|| async {}` in code that *does* pass, since it's not as expressive, and transforming `|| async { ... }` to `async || { ... }` will never make code stop compiling.
```

Confirming: this is accurate, right? There are no cases where moving the `async` could cause code that compiled to stop compiling?

Reply:
As far as I'm aware, yes, except there's one caveat, which is closure signature inference (which we don't guarantee much about at the moment).

We **do** try our best to infer an async closure's signature even if users pass them to functions bounded by "old style" `Fn`+`Future` trait bounds, which look like:

```rust
fn not_an_async_closure<F, Fut, T>(callable: F)
where
    F: FnOnce(i32, i32) -> Fut,
    Fut: Future<Output = T>,
{}
```

We would actually infer the signature to be `async |_: i32, _: i32| -> T` in the above function just fine, even though it's not written with an `async FnOnce` bound but an "old stlye" bound.

**But** there's some corner cases when that second bound isn't a `Future` bound, but a some *other* trait that has a blanket implementation for `T: Future`. I saw this recently with the [`TryFuture`](https://docs.rs/futures/latest/futures/future/trait.TryFuture.html#impl-TryFuture-for-F) which [can't have a `Future<Output = T>` supertrait bound for MSRV reasons](https://github.com/rust-lang/futures-rs/pull/2763)(!?).

I'm going to lightly push back on guaranteeing **all** compatibility in behavior in RFC text between `|| async {}` and `async || {}` because it may force us to unnecessarily hinder the expressivity of async closures for "backwards compatibility" purposes, when the backwards mostly regards calls across the interface of "pre-async closures" and "post-async closures" code.

However, the spirit of my work has been to try to make these as intercompatible as possible without affecting or overly complicating the compiler implementation though, and I'll be certain to note any of these remaining caveats in the async closures stabilization report when that comes around.

I'm confident that in the long tail the ecosystem will adopt `async Fn` (or `AsyncFn`) trait bounds in their `where` clauses, and the callers/the compiler will not end up not be needing to span the gap between async closures and library code written with the "old style" `F: impl Fn() -> Fut, Fut: Future<Output = T>` bounds so much as time goes on, so I'm not particularly concerned about this, tbh.

Reply:
@compiler-errors Thanks for the explanation. Given that it isn't fully guaranteed right now, we shouldn't add this text. (We can evaluate that further if and when someone proposes the lint.)

Can you please add an item to the "unresolved questions" section to capture your mention of "I'll be sure to note any of these remaining caveats in the async closures stabilization report"? That way we'll note the need for that in the tracking issue.

Reply:
Noted